toolchain/semantics/semantics_context.cpp

```diff
@@ -379,6 +379,44 @@ auto SemanticsContext::ImplicitAsImpl(SemanticsNodeId value_id,
     // Type doesn't need to change.
     return ImplicitAsKind::Identical;
   }
+
+  auto array_type = semantics_ir_->GetTypeAllowBuiltinTypes(as_type_id);
+  auto array_node = semantics_ir_->GetNode(array_type);
+  if (array_node.kind() == SemanticsNodeKind::ArrayType) {
+    auto [bound_node_id, element_type_id] = array_node.GetAsArrayType();
+    if (semantics_ir_
+            ->GetNode(semantics_ir_->GetTypeAllowBuiltinTypes(value_type_id))
+            .kind() == SemanticsNodeKind::TupleType) {
+      auto tuple_type_block_id =
+          semantics_ir_
+              ->GetNode(semantics_ir_->GetTypeAllowBuiltinTypes(value_type_id))
+              .GetAsTupleType();
+      auto type_block = semantics_ir_->GetTypeBlock(tuple_type_block_id);
+      if (type_block.size() == semantics_ir_->GetArrayBound(bound_node_id)) {
+        for (auto type : type_block) {
+          if (type != element_type_id) {
+            if (output_value_id != nullptr) {
+              *output_value_id = SemanticsNodeId::BuiltinError;
+            }
+            return ImplicitAsKind::Incompatible;
+          }
+        }
+        auto array_value_type = CanonicalizeType(array_type);
+        AddNode(SemanticsNode::ArrayValue::Make(value.parse_node(),
+                                                array_value_type, value_id));
+        if (output_value_id != nullptr) {
+          *output_value_id =
+              semantics_ir_->GetTypeAllowBuiltinTypes(array_value_type);
+        }
+        return ImplicitAsKind::Compatible;
+      }
```

Comment:
This can be made a bit simpler with a standard library algorithm:

```suggestion
      if (type_block.size() == semantics_ir_->GetArrayBound(bound_node_id) &&
          std::all_of(type_block.begin(), type_block.end(),
                      [&] (auto type) { return type == element_type_id; })) {
        if (output_value_id != nullptr) {
          *output_value_id =
              AddNode(SemanticsNode::ArrayValue::Make(value.parse_node(),
                                                      as_type_id, value_id));
        }
        return ImplicitAsKind::Compatible;
      }
```

A couple of other things here:

- We don't need to `CanonicalizeType(array_type)`, because we already have the `SemanticsTypeId` of the array type -- that's `as_type_id`.
- We should be filling in the `*output_value_id` with the array value node itself, not the type of that node. And it looks like we're not supposed to actually build the node unless an `output_value_id` is provided.